packages/core/src/types.ts

```diff
@@ -140,7 +140,7 @@ export type BaseAction<
   TAction extends BaseActionObject
 > =
   | SimpleActionsOf<TAction>['type']
-  | TAction
+  | (string extends TAction['type'] ? never : TAction)
```

Comment:
This is losing us some functionality when `TAction === BaseActionObject` (so... most of the time) as random **objects** won't be allowed anymore. Since this wasn't really type-safe anyway and that feature is little known - I'm OK with this.

Reply:
Can you show what kind of code would break (type-wise) with this change?

Reply:
Take a look here:
https://github.com/statelyai/xstate/blob/11d815ee92044e0fb8f4ce89a6e6b2f6cf8dd2b0/packages/core/test/actions.test.ts#L903-L907

In here, with this PR, we still allow `entry: 'definedAction'` but we don't allow `entry: { type: 'definedAction' }`. That is because this action is not really "defined" through `TAction`. Both flavors of defining actions are "unsafe" in the very same manner - but the string variant is very common and it doesn't collide with other builtin action objects (like `assign` or whatever) because a string and an object don't share any structure/property with each other. 

I believe that with typegen the "object variant" isn't even supported and we derive "TAction" only from the "string variant". Although we probably don't literally derive the `TAction`, it's just derived conceptually. This somewhat raises a question for the future - how one should define their actions? If actually defining `schema.actions` - should one be required to define there even "simple actions" (string only) or should they only need to define complex ones there and we should construct the final `TAction` based on `schema.actions` and the derived typegen data?

I think this issue here would also be fixed if we add "callable signatures" to our builtin action objects because then those "object variants" wouldn't match our bultin action objects. This is something I want to experiment with - but it's a big task and I'm not sure if I don't end up with some weird TS problems that might block this work. So given that... it's not certain when that other solution for the problem that is being fixed here can be delivered.

Another idea that just popped up to my mind... we could also try introducing brands to our action objects and maybe use something like `TAction & { __brand?: never }`. I wonder though - what's our end goal for managing/typing actions. I think that it makes sense to either:
- require everything to be declared up front
- only require "complex" ones to be declared and infer the rest

The first solution feels quite strict but OTOH this could be automated pretty neatly by the IDE extension and we wouldn't end up with magic rules that might be hard to grok for newcomers.

Reply:
It sounds like we should make an RFC for this. We still want to allow this:

```js
entry: [
  'definedAction', // simple action without parameters - allowed
  { type: 'greetAction', message: 'hello' }, // action with parameters - allowed
  'undefinedAction' // this should cause a type error - disallowed
]
```